src/main/java/com/google/devtools/build/lib/actions/PathMapper.java

```diff
@@ -33,6 +35,43 @@
  * part (e.g. "k8-fastbuild") from exec paths to allow for cross-configuration cache hits.
  */
 public interface PathMapper {
+  /**
+   * Retrieve the {@link PathMapper} instance stored in the given {@link StarlarkSemantics} via
+   * {@link #storeIn(StarlarkSemantics)}.
+   */
+  static PathMapper loadFrom(StarlarkSemantics semantics) {
+    return semantics.get(SEMANTICS_KEY);
+  }
+
+  /**
+   * Creates a new {@link StarlarkSemantics} instance which causes all Starlark threads using it to
+   * automatically apply this {@link PathMapper} to all struct fields of {@link
+   * com.google.devtools.build.lib.starlarkbuildapi.FileApi}.
+   *
+   * <p>This is meant to be used when evaluating user-defined callbacks to Starlark variants of
+   * custom command lines that are evaluated during the execution phase.
+   *
+   * <p>Since any unmapped path appearing in a command line will prevent cross-configuration cache
+   * hits, this mapping is applied automatically instead of requiring users to explicitly map all
+   * paths themselves. As an added benefit, this allows actions to opt into path mapping without
+   * actual changes to their command line code.
+   */
+  @CheckReturnValue
+  default StarlarkSemantics storeIn(StarlarkSemantics semantics) {
+    return new StarlarkSemantics(semantics.toBuilder().set(SEMANTICS_KEY, this).build()) {
```

Comment:
is it possible to avoid creating new instance of StarlarkSemantics for NOOP case?

Reply:
I can do that, but could you confirm that the benchmarks are good before we make that change?

Reply:
The benchmark is still running. I will let you know asap

Reply:
Here is the result 

```
metric        median              Δ                  1-pval         
   cpu: 16852.300s  ±505.3s  +442.1s, +2.7% 0.92 (weakly significant)
memory:     11110MB   ±1.6MB  -2.0MB, -0.0% 0.13 (not significant)   
system:  1078.910s   ±55.8s    -5.8s, -0.5% 0.13 (not significant)   
  wall:  1402.026s   ±53.6s    +3.5s, +0.3% 0.13 (not significant)  
```

Wondering if preventing new instances of StarlarkSemantics for NOOP case would help reduce the regression even further?

Reply:
Do you have a flamegraph that can attribute the additional CPU time? We can try with the default semantics, but if that avoids the regression, we should also be able to do better for the path mapping case (we may have missed another cache).